### PR TITLE
Neural Net blog

### PR DESCRIPTION
**Pull request by lincent**
things to update
* all images to be recreated
* additional graphs demoing linear function combination v with activation functions
* updated graphs
* maybe some resources to go read

### CODE REVIEW

File: neural_net_basics/assets/architecture.png
Comment (chrisprice) [inline neural_net_basics/assets/architecture.png:1]: Please can you cite any images that you've /acquired/ or create your own versions of them.

File: neural_net_basics/neural-network.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+## Intro
+Recently I've been learning about Neural Networks and found myself wanting to write about the how they work from a basic level.
+
+## What is a neural network made of?
+
+All neural networks are made up of a network of neurons, much like the brain that they model, but what is a neuron?
+
+A neuron is made up of it's inputs, some weights, a bias and activation function and an output and was originally intended to replicate the biological neuron.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: Please can you run the whole piece through a spelling/grammar check. Specific things in this sentence that caught my eye (`it's`, `a bias and activation function and an output and was...`)

File: neural_net_basics/neural-network.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+* $b$ is the bias
+* $n$ is the number of inputs
+
+
+![png](/neural_net_basics//assets//neuron.png)
+
+## Why is an activation function important?
+A neuron, with a single input, is a linear function, $y = ax+b$, and as such can only form simple decision boundaries. An increase in the number of inputs and neurons in the neural network will still result in a linear function. This is because multiple combined linear functions always results in a linear function. 
+
+To resolve this issue, and allow the network to find complex decision boundaries, one that wiggles around separating multiple classes, we need to use activation functions, they add non-linearity.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md:26]: I understand what you mean but as the audience for this post seems to be beginners, I think a diagram for each paragraph would really help with understandability.

File: neural_net_basics/neural-network.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+### Logistic Function a.k.a Sigmoid function
+![png](/neural_net_basics//assets//sigmoid.png)
+$$ f(x) = \frac{1}{1 + e^{-x}}$$
+The Logistic function, a variant of the sigmoid function, initially replaced the step function. Like the Step function it limits the output of a neuron between 0 and 1, but doesn't have the same issues around the zero mark. This makes it easier to train as a change to the weights and biases will always have an effect, except at the extremes.
+
+### Rectified Linear Unit (ReLU)
+![png](/neural_net_basics//assets//relu.png)
+$$ f(x) = max(0,x) $$
+The Rectified Linear  (ReLU) has become the most popular and in most cases the default activation function to use for deep neural networks. It takes the maximum value of either 0 or the input, essentially putting any negative input value to 0 immediately.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: Think you've dropped Unit
```suggestion
The Rectified Linear Unit (ReLU) has become the most popular and in most cases the default activation function to use for deep neural networks. It takes the maximum value of either 0 or the input, essentially putting any negative input value to 0 immediately.
```

File: neural_net_basics/neural-network.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+### SoftMax
+$$ \delta(z) = \frac{e^{z_{i}}}{\sum_{j=1}^{K}e^{z_{j}}} $$
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: It might be worth pointing out that SoftMax is dissimilar to the others in that it takes a vector

File: neural_net_basics/neural-network.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+1. Input Layer
+2. Hidden Layers
+3. Output Layer
+
+### 1. Input Layer
+The input layer is purely the values of the input data, no calculations take place. There is a neuron for each input value, for image classification neural networks, for grey scale images, there would be a node for each pixel. 
+
+### 2. Hidden Layers
+If there is a single hidden layer then the neural network is considered to be a shallow neural network. If there are two or more hidden layers then it is considered to be a deep learning neural network.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: This doesn't describe what a hidden layer is.

File: neural_net_basics/neural-network.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+In neural networks, the neurons can be arranged in a variety of ways. For the basic neural network the neurons are arranged into several distinct layers. 
+
+There are three layer categories
+
+1. Input Layer
+2. Hidden Layers
+3. Output Layer
+
+### 1. Input Layer
+The input layer is purely the values of the input data, no calculations take place. There is a neuron for each input value, for image classification neural networks, for grey scale images, there would be a node for each pixel.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: What's the difference between a node and a neuron? If they are distinct, it would be good to describe the difference. If they are synonyms, it aids understanding to always use the canonical term.

File: neural_net_basics/neural-network.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+### 3. Output Layer
+The output layer gives the output of the neural network. For image classification there would be a node for each class of item to be classified, for example in hand written digit recognition there would be 10 nodes, one for each of the 10 digits.
+
+### How the layers are connected
+
+For the basic layered neural network, each neuron takes the outputs of every neuron in the previous layer as it’s input. With the first hidden layer taking the inputs to the neural network directly.
+
+![png](/neural_net_basics//assets//architecture.png)
+
+For a classifier neural network, the output layer has neuron that relates to specific classes of output. For example in a neural network that classifies numerical digits, of which there are 10, you would have 10 output neurons. The neuron with the highest output signifies which class the network has identified.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: This is the same example used a couple of paragraphs up.

File: neural_net_basics/neural-network.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+For the basic layered neural network, each neuron takes the outputs of every neuron in the previous layer as it’s input. With the first hidden layer taking the inputs to the neural network directly.
+
+![png](/neural_net_basics//assets//architecture.png)
+
+For a classifier neural network, the output layer has neuron that relates to specific classes of output. For example in a neural network that classifies numerical digits, of which there are 10, you would have 10 output neurons. The neuron with the highest output signifies which class the network has identified. 
+
+Using the SoftMax algorithm as the activation function for the final layer of a neural network you can convert the values of each classes output into a percentage value, indicating the certainty of the network at its classification.
+
+It has been proven that a neural work which has two layers and uses non-linear activation functions can be a universal function approximator, that can find any mathematical function.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]:
```suggestion
It has been proven that a neural network which has two layers and uses non-linear activation functions can be a universal function approximator, that can find any mathematical function.
```

File: neural_net_basics/neural-network.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+## Determining the weights and biases (the parameters)
+
+So how do the weights and biases of each neuron get set? 
+
+The weights and Biases could be manually updated until the output of the neural network is correct. However neural networks can consist of millions/billions of parameters, so an automated process is required.
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]:
```suggestion
The weights and biases could be manually updated until the output of the neural network is correct. However neural networks can consist of millions/billions of parameters, so an automated process is required. 
```

File: neural_net_basics/neural-network.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+$$ w_{i+1} = w_i - \alpha\frac{\delta L}{\delta w_i}$$
+
+* $w_{i+1}$ is the new weight
+* $w_i$ the current weight
+* $\alpha$ the learning rate
+* $\frac{\delta L}{\delta w_i}$ is the derivative of the change in loss against weight
+
+The learning rate is used to control how much of a change to the weight is made, if the size of the weight change is too big then loss minima may be stepped over or it may never settle in the "valley" of the minima and constantly jump from side to side. If it's too small then it will take longer to reach the local minima.
+
+### Back Propagation - remove?!
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: I'd keep it. I think it's fine to be honest and say something along the lines of it's a complicated subject to explain briefly, check out these resources to learn more.

File: neural_net_basics/neural-network.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+A calculus concept called Chain Rule is used as part of the back propagation to calculate the gradient of the error function with respect to each layers inputs, weights and biases.
+
+To learn more about Gradient Descent and back propagation and how that works, the following resources should give you a good understanding of how it works.
+
+* resource 1
+* resource 2
+* resource 3
+
+### Learning
```
Comment (chrisprice) [inline neural_net_basics/neural-network.md]: ?

Comment (chrisprice) [reply neural_net_basics/neural-network.md]: You need a conclusion to ease the reader's landing back into the real world. Random ideas for a conclusion, pick and choose as you see fit -
We set out to do this, it worked as/not as we expected. It was more/less interesting/challenging/fun that expected. If I did it again I would do/not do.... In doing this work I came across this other area which I'm interested in investigating.

**Review by chrisprice — CHANGES_REQUESTED** [review]

File: neural_net_basics/assets/architecture.png
Comment (lincent) [reply neural_net_basics/assets/architecture.png:1]: yeah I am planning to create my own

File: neural_net_basics/neural-network.md (hunk 1 above)
Comment (lincent) [reply neural_net_basics/neural-network.md]: I think I've caught everything

File: neural_net_basics/neural-network.md (hunk 7 above)
Comment (lincent) [reply neural_net_basics/neural-network.md]: I've removed it and simplified this section
